bindings/cpu_profiler.cc

```diff
@@ -52,6 +57,173 @@ v8::CpuProfilingLoggingMode GetLoggingMode() {
+  {
+    uv_timer_init(loop, &timer);
+    timer.data = this;
+  };
+
+  static void callback(uv_timer_t* handle);
+  void add_listener(const char* profile_id, std::function<void(uv_timer_t*, uint64_t, v8::HeapStatistics&)> cb);
+  void remove_listener(const char* profile_id, std::function<void(uv_timer_t*, uint64_t, v8::HeapStatistics&)>& cb);
+};
+
+void HeapStatisticsTimer::callback(uv_timer_t* handle) {
+  HeapStatisticsTimer* self = static_cast<HeapStatisticsTimer*>(handle->data);
+  self->isolate->GetHeapStatistics(&self->heap_stats);
+  uint64_t ts = uv_hrtime();
+
+  for (auto& listener : self->listeners) {
+    listener.second(handle, ts, self->heap_stats);
+  }
+}
+
+void HeapStatisticsTimer::add_listener(const char* profile_id, std::function<void(uv_timer_t*, uint64_t, v8::HeapStatistics&)> cb) {
+  listeners.emplace(std::string(profile_id), cb);
+
+  if (listeners.size() == 1) {
+    uv_timer_set_repeat(&timer, period_ms);
+    uv_timer_start(&timer, callback, 0, period_ms);
+  }
+}
+
+void HeapStatisticsTimer::remove_listener(const char* profile_id, std::function<void(uv_timer_t*, uint64_t, v8::HeapStatistics&)>& cb) {
+  listeners.erase(std::string(profile_id));
+
+  if (listeners.size() == 0) {
+    uv_timer_stop(&timer);
+  }
+};
+
+class Profiler {
+public:
+  std::unordered_map<std::string, SentryProfile*> active_profiles;
+
+  HeapStatisticsTimer heap_statistics_timer;
+  v8::CpuProfiler* cpu_profiler;
+
+  explicit Profiler(const napi_env& env, v8::Isolate* isolate) :
+    heap_statistics_timer(uv_default_loop()),
+    cpu_profiler(
+      v8::CpuProfiler::New(isolate, kNamingMode, GetLoggingMode())) {
+    napi_add_env_cleanup_hook(env, DeleteInstance, this);
+  }
+
+  static void DeleteInstance(void* data);
+};
+
+void Profiler::DeleteInstance(void* data) {
+  Profiler* profiler = static_cast<Profiler*>(data);
+  profiler->cpu_profiler->Dispose();
+  delete profiler;
+}
+
+class SentryProfile {
+private:
+  uint64_t started_at = 0;
+  uint16_t heap_measurement_index = 0;
+  std::vector<uint64_t> heap_stats_ts;
+  std::vector<uint64_t> heap_stats_usage;
+  std::function<void(uv_timer_t*, uint64_t, v8::HeapStatistics&)> memory_sampler_cb;
+  const char* id;
+  ProfileStatus status;
+
+public:
+  explicit SentryProfile(const char* id) :
+    started_at(uv_hrtime()),
+    heap_measurement_index(0),
+    heap_stats_ts(300),
+    heap_stats_usage(300),
+    memory_sampler_cb([this](uv_timer_t* timer, uint64_t ts, v8::HeapStatistics& stats) {
+    heap_stats_ts.insert(heap_stats_ts.begin() + heap_measurement_index, ts - started_at);
+    heap_stats_usage.insert(heap_stats_usage.begin() + heap_measurement_index, static_cast<uint64_t>(stats.used_heap_size()));
+    ++heap_measurement_index;
+      }), id(id), status(ProfileStatus::kNotStarted) {}
+
+
+      const std::vector<uint64_t>& heap_usage_timestamps();
+      const std::vector<uint64_t>& heap_usage_values();
+      const uint16_t& heap_usage_entries_count();
```

Comment:
there should be another `const` at the end of the function signature (e.g. `const std::vector<uint64_t>& heap_usage_values() const`. at the beginning you're saying the return type (e.g. `std::vector`) is `const`, at the end you're saying the function does not mutate state